feat: recompute tipset to generate missing events if event indexing is enabled

aarshkshah1992 opened the pull request:
For https://github.com/filecoin-project/lotus/issues/12453.

If we're unable to find events for a given receipt, we recompute the tipset to generate events and re-read the events true to enable auto-repair for events. This will work during reconciliation as well.

rvagg commented on chain/index/indexer.go:
```diff
@@ -21,6 +21,7 @@ var _ Indexer = (*SqliteIndexer)(nil)
 
 // IdToRobustAddrFunc is a function type that resolves an actor ID to a robust address
 type IdToRobustAddrFunc func(ctx context.Context, emitter abi.ActorID, ts *types.TipSet) (address.Address, bool)
+type tipsetExecutorFnc func(ctx context.Context, ts *types.TipSet) (cid.Cid, cid.Cid, error)
```
Leaky abstraction here, since the two CIDs are always discarded. A couple of options to make this a bit nicer:

1. Just use an `error` return and make the wrapper function 1 line longer to discard the others (even though currently it doesn't even need to be a wrapper, just `indexer.SetTipsetExecutorFnc(sm.RecomputeTipSetState)` would do).
2. Just call it `recomputeTipSetState` and admit that this is all we're expecting here.
3. Make `SqliteIndexer` depend on `StateManager` and load it in via fx to avoid jumping through hoops and to be completely honest.


(also, `Fnc` -> `Func`?)

aarshkshah1992 replied:
@rvagg Fixed this by using a  `recomputeTipSetState` function here that simply returns an error and passing in a wrapper to hide the AMT cids during construction.

rvagg commented:
I'm OK with this in principle since the only path to calling this would be from `backfillIndex` -> `indexTipsetWithParentEvents`. Getting there via `Apply` should not encounter it because the amt should be there.

Aside from the review above, the only other comment I have is that it was kind of difficult tracing the calls with a `git grep` down into this because it jumps around through 3 files to get to the top calls, that makes me a little sus about how clean the abstractions are, but I don't have any good suggestions, just an observation.

aarshkshah1992 commented:
@rvagg I have addressed your review. I agree that there is some indirection involved in tracing the injection -> use flow for the repair function but can't think of a quick cleaner way here too.

aarshkshah1992 commented:
@rvagg Please let me know if I have your ✔️ on this PR and I can merge it to the parent branch.
